### PR TITLE
File baseline reduce intent bug and work around it in parallel array.find() implementation

### PR DESCRIPTION
**Pull request by bradcray**
This captures a bug that showed up in my new array.find() implementation for arrays indexed by bools.  Specifically, something about reduce intents over bools on forall loops, in combination to uses of that bool within the loop's body, causes an internal error that I don't see in similar patterns for reduce intents over ints or other types.

### CODE REVIEW

**Comment by bradcray** [comment]
@e-kayrakli : This may be of interest to you since you reviewed the parallel array.find() routine.
@vasslitvinov : It may be of interest to you since it seems at least somewhat related to reduce intents (though it's not obvious that those are at fault specifically... it could be something else that they happen to use).

File: test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl [hunk 1]
```diff
@@ -0,0 +1,8 @@
+var A: [false..true] real = [1.2, 3.4];
+
+var locIdx = max(bool);
+forall i in A.domain with (min reduce locIdx) {
+  if (i < locIdx) {
+  }
```
Comment (e-kayrakli) [inline test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl:6]: did you want to set `locIdx` here?

File: modules/internal/ChapelArray.chpl [hunk 2]
```diff
@@ -1831,7 +1831,7 @@ module ChapelArray {
         return false;
 
       // workaround for #21749:
-      } else if __primitive("call and fn resolves", "max", fullIdxType) {
+      } else if this.idxType != bool && __primitive("call and fn resolves", "max", fullIdxType) {
```
Comment (e-kayrakli) [inline modules/internal/ChapelArray.chpl:1834]: maybe expand the comment above because now this line has 2 workarounds :)

**Review by e-kayrakli — APPROVED** [review]

File: test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl (hunk 1 above)
Comment (bradcray) [reply test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl:6]: It doesn't need to be to hit the internal error, and I was trying to keep the reproducer as minimal as possible.

File: modules/internal/ChapelArray.chpl (hunk 2 above)
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:1834]: Now I know why I don't usually add comments!  :D

File: test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl (hunk 1 above)
Comment (e-kayrakli) [reply test/parallel/forall/reduce-intents/ri-bool-baseline-bug.chpl:6]: sounds good

File: modules/internal/ChapelArray.chpl (hunk 2 above)
Comment (e-kayrakli) [reply modules/internal/ChapelArray.chpl:1834]: Whether comments increase or decrease maintainability is definitely an interesting discussion :D

Comment (bradcray) [reply modules/internal/ChapelArray.chpl:1834]: Worst of all, I added this one without anyone even requesting me to!  Totally out of character... :)

Comment (bradcray) [reply modules/internal/ChapelArray.chpl:1842]: OK, there's a much bigger comment there now, for better or worse

**Comment by cassella** [comment]
I'm all caught up on main and freshly recompiled, and the future passes for me.

```
Future (bug: --baseline problem w/ bool reduce intents and uses of bool within forall) [Success matching program output for parallel/forall/reduce-intents/ri-bool-baseline-bug]
[Summary: #Successes = 0 | #Failures = 0 | #Futures = 1 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 1 ]
```

```
chpl version 1.30.0 pre-release (a78e45a929)
  built with LLVM version 14.0.0

CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: none
CHPL_LLVM: system
CHPL_AUX_FILESYS: none

gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

Ubuntu clang version 14.0.0-1ubuntu1
```

**Comment by bradcray** [comment]
Hi Paul — It turns out that a GASNet (and presumably any `--no-local` compile) doesn't trigger the same internal error.  I unfortunately didn't see this in time to narrow the scope of the future last night, but our nightly testing came to the same conclusion.

**Pull request merged**